doc/reference/trd-appid.md

```diff
@@ -0,0 +1,651 @@
+requirements of a particular Tock device use case. An application can
+consist of multiple process binaries to support versioning.
+
+**Application binary**: a code image compiled to run in a Tock
+process.
+
+**Application identifier**: a numerical identifier for an application.
+Every process binary has a single application identifier, and multiple
+process binaries can share the same application identifier. An
+application identifier can be persistent across boots or restarts of a
+process binary. 
+
+**Application credentials**: data that binds an application identifier
+to an application binary. Application credentials are usually stored
+in [Tock Binary Format][TBF] footers.
+
+**Verifier**: a component of the Tock kernel which is responsible for
+validating application credentials and assigning application
+identifiers. 
+
+**Verifier Policy**: the algorithm that a Verifier uses to assign
+application identifiers to a process binary. A verifier policy defines
+a application identifier space. Tock kernels using different verifier
+policies may and often do assign different application identifiers for
+the same process binary.
+
+**Global application identifier**: an application identifier which is
+globally unique for the verifier policy that assigned it. All
+instances of the application loaded with that verifier policy have
+this identifier and no instances of other applications loaded with
+that Verifier policy have this identifier.
+
+**Local application identifier**: an application identifier which is
+locally unique for the verifier policy that assigned it.  No other
+process binary on the same Tock device have this identifier.
+
+In normal use of Tock, process binaries are copied into an application
+flash region by a software tool. When the Tock kernel boots, it scans
+this application flash region for process binaries. After inspecting the
+application binary and TBF headers in a process binary, the kernel
+decides whether to load it into a process and run it.
+
+3 Application identifiers and application credentials
+===============================
+
+There is a relationship between application identifiers and
+credentials, but they are not the same thing. An application
+identifier is a numerical representation of the application's
+identity, while credentials are the data that, combined with a
+verifier policy, bind identifiers to binaries. Suppose there are two
+versions (v1.1 and v1.2) of the same application. They have different
+binaries. Their application credentials consist of a cryptographic
+hash of their binary signed by a known public key. The public key
+defines the application identifier: all versions of this application
+have credentials signed by this key.  The two versions have different
+credentials, because their hashes differ, but they have the same
+application identifier.
+
+Every running Tock process MUST have an application identifier.
+Application identifiers MUST be unique across running processes in a
```

Comment:
I don't want to bind this to specific values of the enum because, well, they might change or be renamed.  But I agree that we need a more precise statement of the process state machine and what "running" means.

In terms of the current enum, it means Running, Yielded, StoppedRunning, StoppedYielded.